WIP: scheduler_perf: fix shutdown and increase timeout

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This is part of the investigations of https://github.com/kubernetes/kubernetes/issues/125223.

Manually running the benchmark showed that it didn't shut down as quickly as it should have (fixed by first commit) and looking at the CI logs revealed timeouts (perhaps fixed by second commit).

#### Which issue(s) this PR fixes:
Related-to: 125223

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
scheduler_perf: shut down apiserver clients before apiserver

The cancellation of the context happened after the cleanup of the apiserver, so
clients using that context were kept running. That wasn't the intent and causes
a slow shutdown because the apiserver delays its shutdown when it has active
clients.

The fix is to create a new cancellation context and to use that for the
clients. The automatic cancellation of it then happens before the apiserver
cleanup.

pohly committed:
scheduler_perf: increase per-test timeout

There were some "context canceled" errors in the
ci-benchmark-scheduler-perf-master job, but no obvious big performance
regression when comparing v1.30.0 and current master (~ v1.31.0 alpha
1). Perhaps a small increase is necessary and sufficient.

kerthcet commented:
So this is just for investigations and will be reverted in the future, right?

kerthcet commented on test/integration/scheduler_perf/scheduler_perf.go:
```diff
@@ -756,8 +756,8 @@ func RunBenchmarkPerfScheduling(b *testing.B, outOfTreePluginRegistry frameworkr
 					framework.StartEtcd(b, output)
 
-					// 30 minutes should be plenty enough even for the 5000-node tests.
-					timeout := 30 * time.Minute
```
I checked the scheduler_perf log, I think the root cause is we're timeout waiting for the pod to be scheduled, code here:
https://github.com/kubernetes/kubernetes/blob/9d8edcac2f3d2b573d0f32a20f2e51cb5268907d/test/integration/scheduler_perf/scheduler_perf.go#L1309

Also see the error log:
    `scheduler_perf.go:1017: FATAL ERROR: op 9: error in waiting for pods to get scheduled: at least pod test/test-drah9mnh is not scheduled: context deadline exceeded`

I do check the time, it tells the truth, and we stuck in this ` scheduler_perf.go:1323: namespace: test, pods: want 1000, got 996` for almost 10 mins, something wrong here, the last 4 pods can't be scheduled.

So I think extending the time doesn't work.

kerthcet replied:
Maybe https://github.com/kubernetes/kubernetes/pull/125360 can help with this.

pohly replied:
If pods get moved to the backoff queue, they can get stuck for a long time until the retry timeout happens.

I can't reproduce that locally and it worked before, so bumping up the timeout is just a way to check whether it just became a little bit slower. We can revert if it doesn't help.

pohly replied:
I'm still waiting for https://github.com/kubernetes/test-infra/pull/32707 to get merged. That would be a way to test in a pre-submit.

pohly replied:
/test pull-scheduler-perf

pohly replied:
@kerthcet: you are right, extending the timeout won't help.

I now have a log at -v5: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/125313/pull-kubernetes-scheduler-perf/1801955385091821568/build-log.txt

The failure there is:
```
    scheduler_perf.go:1017: FATAL ERROR: op 7: error in waiting for pods to get scheduled: at least pod test/test-dra8sj4s is not scheduled: context deadline exceeded
```

Searching for that pod and the associated ResourceClaim shows:
```console
$ grep -e test/test-dra8sj4s -e test/test-dra8sj4s-resource-zxm9r build-log.txt 
    graph_builder.go:692: I0615 12:34:14.013092] GraphBuilder process object apiVersion="v1" kind="Pod" object="test/test-dra8sj4s" uid="99b17925-47e6-4afc-9935-2f9685f7abc8" eventType="add" virtual=    <bool>: false
    eventhandlers.go:128: I0615 12:34:14.018764] Add event for unscheduled pod pod="test/test-dra8sj4s"
I0615 12:34:14.018781   27079 dynamicresources.go:474] "pod unschedulable" reason="pod \"test/test-dra8sj4s\": ResourceClaim not created yet"
I0615 12:34:14.018790   27079 scheduling_queue.go:539] "Status after running PreEnqueue plugin" pod="test/test-dra8sj4s" plugin="DynamicResources" status="pod \"test/test-dra8sj4s\": ResourceClaim not created yet"
    controller.go:557: I0615 12:34:19.163726] checking pod="test/test-dra8sj4s" podClaim="resource"
    controller.go:660: I0615 12:34:19.171506] Created ResourceClaim podClaim="resource" claim="test/test-dra8sj4s-resource-zxm9r" pod="test/test-dra8sj4s"
    graph_builder.go:692: I0615 12:34:19.180410] GraphBuilder process object apiVersion="v1" kind="Pod" object="test/test-dra8sj4s" uid="99b17925-47e6-4afc-9935-2f9685f7abc8" eventType="update" virtual=    <bool>: false
    eventhandlers.go:151: I0615 12:34:19.181285] Update event for unscheduled pod pod="test/test-dra8sj4s"
I0615 12:34:19.181378   27079 scheduling_queue.go:539] "Status after running PreEnqueue plugin" pod="test/test-dra8sj4s" plugin="DynamicResources" status="resourceclaim.resource.k8s.io \"test-dra8sj4s-resource-zxm9r\" not found"
    controller.go:509: I0615 12:34:19.184186] nothing to do for pod, scheduler will deal with it pod="test/test-dra8sj4s"
    graph_builder.go:692: I0615 12:34:19.188535] GraphBuilder process object apiVersion="resource.k8s.io/v1alpha2" kind="ResourceClaim" object="test/test-dra8sj4s-resource-zxm9r" uid="36380712-b9fb-4c38-a264-9e61e08d28e7" eventType="add" virtual=    <bool>: false
    controller.go:304: I0615 12:34:19.190391] test-driver.cdi.k8s.io/resource controller: Adding new work item key="claim:test/test-dra8sj4s-resource-zxm9r"
    controller.go:557: I0615 12:34:19.948900] checking pod="test/test-dra8sj4s" podClaim="resource"
    controller.go:588: I0615 12:34:19.948940] claim already created pod="test/test-dra8sj4s" podClaim="resource" resourceClaim="test-dra8sj4s-resource-zxm9r"
    controller.go:509: I0615 12:34:19.948966] nothing to do for pod, scheduler will deal with it pod="test/test-dra8sj4s"
    controller.go:841: I0615 12:34:19.949104] claim reserved for counts currentCount=    <int>: 0 claim="test/test-dra8sj4s-resource-zxm9r" updatedCount=    <int>: 0 builtinController=    <bool>: false
    controller.go:375: I0615 12:34:20.649171] test-driver.cdi.k8s.io/resource controller: processing key="claim:test/test-dra8sj4s-resource-zxm9r"
    controller.go:519: I0615 12:34:20.649198] test-driver.cdi.k8s.io/resource controller: ResourceClaim waiting for first consumer key="claim:test/test-dra8sj4s-resource-zxm9r"
    controller.go:379: I0615 12:34:20.649222] test-driver.cdi.k8s.io/resource controller: completed key="claim:test/test-dra8sj4s-resource-zxm9r"
    scheduler_perf.go:1017: FATAL ERROR: op 7: error in waiting for pods to get scheduled: at least pod test/test-dra8sj4s is not scheduled: context deadline exceeded
```

The relevant part is this:
```
    eventhandlers.go:151: I0615 12:34:19.181285] Update event for unscheduled pod pod="test/test-dra8sj4s"
I0615 12:34:19.181378   27079 scheduling_queue.go:539] "Status after running PreEnqueue plugin" pod="test/test-dra8sj4s" plugin="DynamicResources" status="resourceclaim.resource.k8s.io \"test-dra8sj4s-resource-zxm9r\" not found"
```

This could be the race between event handler triggered by informer and the plugin using the assume cache. The fix is in https://github.com/kubernetes/kubernetes/pull/124595. Let's see whether the new job passed there...

alculquicondor commented:
/test pull-scheduler-perf

alculquicondor commented:
Still not working.

pohly commented:
/test pull-perf-tests-scheduler-perf

I had to ask on Slack. Let's see if this works.

pohly commented:
Nope :disappointed:

pohly commented:
I need to move the pull job - will get to it tomorrow.

pohly commented:
/test pull-kubernetes-scheduler-perf

From https://github.com/kubernetes/test-infra/pull/32748

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly committed:
test: filter "go test" output with gotestsum instead of grep

Filtering the output with grep leads to hard to read log output, e.g. from
pull-kubernetes-unit:

    +++ [0613 15:32:48] Running tests without code coverage and with -race
    {"Time":"2024-06-13T15:33:47.845457374Z","Action":"output","Package":"k8s.io/kubernetes/cluster/gce/cos","Test":"TestCreateMasterAuditPolicy","Output":"        /tmp/configure-helper-test47992121/kube-env: line 1: `}'\n"}
    {"Time":"2024-06-13T15:33:49.053732803Z","Action":"output","Package":"k8s.io/kubernetes/cluster/gce/cos","Output":"ok  \tk8s.io/kubernetes/cluster/gce/cos\t2.906s\n"}

We can do better than that. When feeding the output of the "go test" command(s)
into gotestsum *while it runs*, we can use --format=standard-quiet (= normal go
test output) or --format=standard-verbose (= `go test -v`) when FULL_LOG is
requested to get nicer output.

gotestsum doesn't support reading from a pipe. We can work around that with
<( go test ... ) which gives us a file name (/dev/fd/<num> on Linux) for
the `--raw-command cat <file name>`.

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly commented:
I'm going to use this PR for some more experiments...

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly committed:
WIP: experiment

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly committed:
test

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly committed:
disable false warning

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly committed:
WIP: use forked gotestsum

pohly committed:
make test: use gotestsum in pipe

pohly committed:
component-base/logs: inherit defaults from klog

A test might use ktesting.SetDefaultVerbosity to configure different defaults.
Those only have an effect when component-base/logs then honors those defaults
instead of applying its own/

pohly committed:
WIP

pohly committed:
WIP

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly commented:
/test pull-kubernetes-unit
/test pull-kubernetes-integration

pohly committed:
resourceclaim controller: add missing log output

The logging was fairly complete about *not* doing something, but the actual
ResourceClaim creation was not logged.

pohly committed:
WIP

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly committed:
WIP: cherry-pick assume cache (https://github.com/kubernetes/kubernetes/pull/124595)

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly committed:
DRA scheduler: use assume cache to list claims

This finishes the transition to the assume cache as source of truth for the
current set of claims.

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly committed:
WIP: logging

pohly committed:
WIP: shorten timeout

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly committed:
WIP: more logging

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly committed:
revert parts of https://github.com/kubernetes/kubernetes/pull/124618

pohly commented:
/test pull-kubernetes-scheduler-perf

pohly commented:
I moved all of the relevant commits elsewhere, this is not needed anymore.
